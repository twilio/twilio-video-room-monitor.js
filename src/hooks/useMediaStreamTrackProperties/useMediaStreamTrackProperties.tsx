import { useEffect, useState } from 'react';

export default function useMediaStreamTrackProperties(mediaStreamTrack: MediaStreamTrack) {
  const [muted, setMuted] = useState(mediaStreamTrack.muted);
  const [readyState, setReadyState] = useState(mediaStreamTrack.readyState);

  useEffect(() => {
    setMuted(mediaStreamTrack.muted);
    setReadyState(mediaStreamTrack.readyState);
    if (mediaStreamTrack) {
      const handleMuted = () => setMuted(mediaStreamTrack.muted);
      const handleReadyState = () => setReadyState(mediaStreamTrack.readyState);

      mediaStreamTrack.addEventListener('mute', handleMuted);
      mediaStreamTrack.addEventListener('unmute', handleMuted);
      mediaStreamTrack.addEventListener('ended', handleReadyState);
      return () => {
        mediaStreamTrack.removeEventListener('mute', handleMuted);
        mediaStreamTrack.removeEventListener('unmute', handleMuted);
        mediaStreamTrack.removeEventListener('ended', handleReadyState);
      };
    }
  }, [mediaStreamTrack]);

  return {
    muted,
    readyState,
    id: mediaStreamTrack.id,
    label: mediaStreamTrack.label,
    kind: mediaStreamTrack.kind,
  };
}
