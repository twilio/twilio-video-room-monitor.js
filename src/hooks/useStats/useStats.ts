import React from 'react';
import { RoomStatsContext } from '../../components/RoomStatsProvider/RoomStatsProvider';

export default function useStats() {
  const stats = React.useContext(RoomStatsContext);
  return stats;
}
