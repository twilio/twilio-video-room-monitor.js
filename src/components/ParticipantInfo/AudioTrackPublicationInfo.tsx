import React from 'react';
import styled from 'styled-components';
import {
  LocalAudioTrack,
  LocalAudioTrackPublication,
  LocalAudioTrackStats,
  RemoteAudioTrack,
  RemoteAudioTrackPublication,
  RemoteAudioTrackStats,
} from 'twilio-video';
import useIsTrackEnabled from '../../hooks/useIsTrackEnabled/useIsTrackEnabled';
import { useTrackBandwidth, useTrackData } from '../../hooks/useStats/useStats';
import useTrack from '../../hooks/useTrack/useTrack';
import { theme } from '../theme';
import Datum from '../typography/Datum/Datum';

export const AudioTrackInfo: React.FC<{
  track: LocalAudioTrack | RemoteAudioTrack;
  trackSid: string; // Passing trackSid from the publication object because it not on the LocalAudioTrack object
}> = ({ track, trackSid }) => {
  const isEnabled = useIsTrackEnabled(track);
  const trackBandwidth = useTrackBandwidth(trackSid);
  const trackData = useTrackData(trackSid) as LocalAudioTrackStats | RemoteAudioTrackStats | null;

  return (
    <>
      <Datum label="isEnabled" value={String(isEnabled)} />
      <Datum label="Bandwidth" value={String(trackBandwidth) + 'kbps'} />
      {trackData && (
        <>
          <Datum label="Codec" value={String(trackData.codec)} />
          <Datum label="Jitter" value={String(trackData.jitter)} />
          <Datum label="Packets Lost" value={String(trackData.packetsLost)} />
        </>
      )}
    </>
  );
};

const Container = styled.div`
  &:not(:last-child) {
    border-bottom: 1px solid ${theme.borderColor};
    margin-bottom: 3px;
  }
  padding-bottom: 3px;
`;

export const AudioTrackPublicationInfo: React.FC<{
  publication: LocalAudioTrackPublication | RemoteAudioTrackPublication;
}> = ({ publication }) => {
  const track = useTrack(publication) as LocalAudioTrack | RemoteAudioTrack | undefined;

  return (
    <Container>
      <Datum label="Name" value={publication.trackName} />
      <Datum label="SID" value={publication.trackSid} />
      <Datum label="isSubscribed" value={String(!!track)} />
      {track && <AudioTrackInfo track={track} trackSid={publication.trackSid} />}
    </Container>
  );
};
