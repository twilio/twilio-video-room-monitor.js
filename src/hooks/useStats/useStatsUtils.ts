import { StatsReport } from 'twilio-video';
import { sum } from 'd3-array';
import useStats from './useStats';

export function getAllStats(statsReports: StatsReport[]) {
  const initialValue = {
    localAudioTrackStats: [],
    localVideoTrackStats: [],
    remoteAudioTrackStats: [],
    remoteVideoTrackStats: [],
    peerConnectionId: '',
  };

  return statsReports.reduce((p, c) => {
    return {
      localAudioTrackStats: [...p.localAudioTrackStats, ...c.localAudioTrackStats],
      localVideoTrackStats: [...p.localVideoTrackStats, ...c.localVideoTrackStats],
      remoteAudioTrackStats: [...p.remoteAudioTrackStats, ...c.remoteAudioTrackStats],
      remoteVideoTrackStats: [...p.remoteVideoTrackStats, ...c.remoteVideoTrackStats],
      peerConnectionId: '',
    };
  }, initialValue);
}

export function getAllTracks(statsReports: StatsReport[]) {
  const statsReport = getAllStats(statsReports);
  const { localAudioTrackStats, localVideoTrackStats, remoteAudioTrackStats, remoteVideoTrackStats } = statsReport;

  const allTracks = [
    ...localAudioTrackStats,
    ...localVideoTrackStats,
    ...remoteAudioTrackStats,
    ...remoteVideoTrackStats,
  ];

  return allTracks;
}

export function getTrackData(trackSid: string, statsReports: StatsReport[]) {
  const allCurrentTracks = getAllTracks(statsReports);
  return allCurrentTracks.filter((t) => t.trackSid === trackSid);
}

// Returns an array that only contains tracks that correspond to a simulcast layer
// that is currently in use. All "configured" but inactive layers will be filtered out.
export function removeInactiveLayers(
  previousStats: StatsReport[] | undefined,
  stats: StatsReport[] | undefined,
  trackSid: string
) {
  if (!stats || !previousStats) return null;

  const previousTracks = getTrackData(trackSid, previousStats);
  const currentTracks = getTrackData(trackSid, stats);

  const activeLayers = currentTracks.filter((currentTrack) => {
    const previousTrack = previousTracks.find((t) => t.ssrc === currentTrack.ssrc);
    return currentTrack.bytesSent !== previousTrack?.bytesSent;
  });

  return activeLayers;
}

export const round = (num: number) => Math.round((num + Number.EPSILON) * 10) / 10;

// Returns the bandwidth usage for a given track SID in kilobytes per second.
// Returns null when track doesn't exist, or when information is not available.
export function useTrackBandwidth(trackSid: string) {
  const { stats, previousStats } = useStats();
  if (!stats || !previousStats) return null;

  const currentTrackData = getTrackData(trackSid, stats);
  const previousTrackData = getTrackData(trackSid, previousStats);

  if (currentTrackData.length === 0 || previousTrackData.length === 0) {
    return null;
  }

  const currentBytes = sum(currentTrackData.map((d) => d.bytesReceived ?? d.bytesSent ?? 0));
  const previousBytes = sum(previousTrackData.map((d) => d.bytesReceived ?? d.bytesSent ?? 0));

  const currentTime = currentTrackData[0]?.timestamp;
  const previousTime = previousTrackData[0]?.timestamp;

  // Calculate kilobytes per second. The timestamp is in milliseconds.
  return round((currentBytes - previousBytes) / (currentTime - previousTime)) * 8;
}

// Returns the bandwidth usage for all local or remote tracks in kilobytes per second.
// Returns null when information is not available.
export function getTotalBandwidth(
  kind: 'bytesSent' | 'bytesReceived',
  stats: StatsReport[] | undefined,
  previousStats: StatsReport[] | undefined
) {
  if (!stats || !previousStats) return null;

  const currentTrackData = getAllTracks(stats).filter((track) => typeof track[kind] !== 'undefined');
  const previousTrackData = getAllTracks(previousStats).filter((track) => typeof track[kind] !== 'undefined');

  // Calculate the bandwidth consumption for each individual track
  const bandwidthPerTrack = (
    currentTrackData
      .map((currentTrack) => {
        // Find the corresponding track source from the previousTrackData array.
        const prevTrack = previousTrackData.find((t) => t.ssrc === currentTrack.ssrc);

        // If no corresponding track is found (because the track was recently published),
        // then it won't be possible to compute bandwidth usage, so null will be returned.
        if (!prevTrack) return null;

        const currentBytes = currentTrack[kind] ?? null;
        const prevBytes = prevTrack[kind] ?? null;

        if (currentBytes !== null && prevBytes !== null) {
          // Calculate kilobytes per second. The timestamp is in milliseconds.
          return ((currentBytes - prevBytes) / (currentTrack.timestamp - prevTrack.timestamp)) * 8;
        } else {
          return null;
        }
      })
      // Remove any null values
      .filter((t) => t !== null) as number[]
  )
    // Occasionally, the bytes sent or received can be reset to lower values, which produces negative
    // bandwidth results. Here we discard any of these negative results.
    .filter((t) => t >= 0);

  return round(sum(bandwidthPerTrack));
}

export function useTrackData(trackSid: string) {
  const { stats, previousStats } = useStats();
  if (!stats || !previousStats) return null;

  const trackData = removeInactiveLayers(previousStats, stats, trackSid);

  return trackData![0];
}
