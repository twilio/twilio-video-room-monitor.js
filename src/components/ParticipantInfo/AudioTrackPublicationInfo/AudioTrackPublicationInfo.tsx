import React from 'react';
import {
  LocalAudioTrack,
  LocalAudioTrackPublication,
  LocalAudioTrackStats,
  RemoteAudioTrack,
  RemoteAudioTrackPublication,
  RemoteAudioTrackStats,
} from 'twilio-video';
import Datum from '../../typography/common/Datum/Datum';
import MediaStreamTrackInfo from '../../typography/trackInformation/MediaStreamTrackInfo/MediaStreamTrackInfo';
import StatsContainer from '../../typography/common/StatsContainer/StatsContainer';
import useIsTrackSwitchedOff from '../../../hooks/useIsTrackSwitchedOff/useIsTrackSwitchedOff';
import useMediaStreamTrack from '../../../hooks/useMediaStreamTrack/useMediaStreamTrack';
import useTrack from '../../../hooks/useTrack/useTrack';
import useTrackSwitchOffReason from '../../../hooks/useTrackSwitchOffReason/useTrackSwitchOffReason';
import { useTrackBandwidth, useTrackData } from '../../../hooks/useStats/useStatsUtils';

export const AudioTrackInfo: React.FC<{
  track: LocalAudioTrack | RemoteAudioTrack;
  trackSid: string; // Passing trackSid from the publication object because it not on the LocalAudioTrack object
}> = ({ track, trackSid }) => {
  const isSwitchedOff = useIsTrackSwitchedOff(track);
  const trackSwitchOffReason = useTrackSwitchOffReason(track);
  const trackBandwidth = useTrackBandwidth(trackSid);
  const trackData = useTrackData(trackSid) as LocalAudioTrackStats | RemoteAudioTrackStats | null;
  const mediaStreamTrack = useMediaStreamTrack(track);

  let lossPercentage: string | null = '';

  if (trackData) {
    const { packetsReceived, packetsSent, packetsLost } = trackData;

    const totalPackets = packetsReceived ?? packetsSent;

    lossPercentage =
      totalPackets && packetsLost !== null ? `${((packetsLost / totalPackets) * 100).toLocaleString()}%` : null;
  }

  return (
    <>
      {/* Only show isSwitchedOff and trackSwitchOffReason for remote tracks */}
      {isSwitchedOff !== undefined && <Datum label="isSwitchedOff" value={isSwitchedOff} />}
      {trackSwitchOffReason !== undefined && <Datum label="Switch Off Reason" value={trackSwitchOffReason} />}
      <Datum label="Bandwidth" value={trackBandwidth?.toLocaleString() + 'kbps'} />
      {trackData && (
        <>
          <Datum label="Codec" value={trackData.codec} />
          <Datum label="Jitter" value={trackData.jitter} />
          <Datum label="Packets Lost" value={trackData.packetsLost} />
          <Datum label="Packet Loss Percentage" value={lossPercentage} />
        </>
      )}
      <MediaStreamTrackInfo track={mediaStreamTrack} />
    </>
  );
};

export const AudioTrackPublicationInfo: React.FC<{
  publication: LocalAudioTrackPublication | RemoteAudioTrackPublication;
}> = ({ publication }) => {
  const track = useTrack(publication) as LocalAudioTrack | RemoteAudioTrack | undefined;

  return (
    <StatsContainer>
      <Datum label="Name" value={publication.trackName} />
      <Datum label="SID" value={publication.trackSid} />
      <Datum label="isSubscribed" value={!!track} />
      {track && <AudioTrackInfo track={track} trackSid={publication.trackSid} />}
    </StatsContainer>
  );
};
