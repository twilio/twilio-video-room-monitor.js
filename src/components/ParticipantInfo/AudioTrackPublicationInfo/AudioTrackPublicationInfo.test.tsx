import React from 'react';
import { shallow } from 'enzyme';
import { AudioTrackPublicationInfo, AudioTrackInfo } from './AudioTrackPublicationInfo';
import { useTrackBandwidth, useTrackData } from '../../../hooks/useStats/useStatsUtils';
import useMediaStreamTrackProperties from '../../../hooks/useMediaStreamTrackProperties/useMediaStreamTrackProperties';
import useTrack from '../../../hooks/useTrack/useTrack';

jest.mock('../../../hooks/useIsTrackEnabled/useIsTrackEnabled', () => () => true);
jest.mock('../../../hooks/useStats/useStatsUtils');
jest.mock('../../../hooks/useTrack/useTrack');
jest.mock('../../../hooks/useMediaStreamTrackProperties/useMediaStreamTrackProperties');

const mockUseTrackBandwidth = useTrackBandwidth as jest.Mock<any>;
const mockUseTrackData = useTrackData as jest.Mock<any>;
const mockUseTrack = useTrack as jest.Mock<any>;
const mockMediaStreamTrackProperties = useMediaStreamTrackProperties as jest.Mock<any>;

mockUseTrackBandwidth.mockImplementation(() => 1234.56);
mockUseTrackData.mockImplementation(() => ({ codec: 'testCodec', jitter: 2, packetsLost: 7 }));
mockUseTrack.mockImplementation(() => 'testTrack');
mockMediaStreamTrackProperties.mockImplementation(() => ({
  id: 'mockId',
  muted: false,
  kind: 'mockKind',
  label: 'mockLabel',
  readyState: 'mockReadyState',
}));

describe('the AudioTrackInfo component', () => {
  it('should render correctly if an audio track is present', () => {
    const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly if there is no audio track data', () => {
    mockUseTrackData.mockImplementationOnce(() => null);
    const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper.find({ label: 'Codec' }).exists()).toBeFalsy();
  });

  describe('the Packet Loss Percentage', () => {
    it('should display null when packetsLost, packetsReceived and packetsSent are not defined', () => {
      mockUseTrackData.mockImplementationOnce(() => ({
        codec: 'testCodec',
        frameRate: null,
        packetsLost: null,
      }));
      const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
      expect(wrapper.find({ label: 'Packet Loss Percentage' }).prop('value')).toBe(null);
    });

    it('should display a value for the RemoteAudioTrack when both packetsReceived and packetsLost are defined', () => {
      mockUseTrackData.mockImplementationOnce(() => ({
        codec: 'testCodec',
        frameRate: null,
        packetsLost: 7,
        packetsReceived: 100,
      }));
      const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
      expect(wrapper.find({ label: 'Packet Loss Percentage' }).prop('value')).toBe('7%');
    });

    it('should display a value for the LocalAudioTrack when both packetsSent and packetsLost are defined', () => {
      mockUseTrackData.mockImplementationOnce(() => ({
        codec: 'testCodec',
        frameRate: null,
        packetsLost: 30,
        packetsSent: 29448,
      }));
      const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
      expect(wrapper.find({ label: 'Packet Loss Percentage' }).prop('value')).toBe('0.102%');
    });

    it('should display 0% for the RemoteAudioTrack when packetsLost is 0 and packetsReceived is defined', () => {
      mockUseTrackData.mockImplementationOnce(() => ({
        codec: 'testCodec',
        frameRate: null,
        packetsLost: 0,
        packetsReceived: 29448,
      }));
      const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
      expect(wrapper.find({ label: 'Packet Loss Percentage' }).prop('value')).toBe('0%');
    });

    it('should display 0% for the LocalAudioTrack when packetsLost is 0 and packetsSent is defined', () => {
      mockUseTrackData.mockImplementationOnce(() => ({
        codec: 'testCodec',
        frameRate: null,
        packetsLost: 0,
        packetsSent: 2134,
      }));
      const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
      expect(wrapper.find({ label: 'Packet Loss Percentage' }).prop('value')).toBe('0%');
    });

    it('should display null for the RemoteAudioTrack when packetsLost is null and packetsReceived is defined', () => {
      mockUseTrackData.mockImplementationOnce(() => ({
        codec: 'testCodec',
        frameRate: null,
        packetsLost: null,
        packetsReceived: 29448,
      }));
      const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
      expect(wrapper.find({ label: 'Packet Loss Percentage' }).prop('value')).toBe(null);
    });

    it('should display null for the LocalAudioTrack when packetsLost is null and packetsSent is defined', () => {
      mockUseTrackData.mockImplementationOnce(() => ({
        codec: 'testCodec',
        frameRate: null,
        packetsLost: null,
        packetsSent: 2134,
      }));
      const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
      expect(wrapper.find({ label: 'Packet Loss Percentage' }).prop('value')).toBe(null);
    });
  });
});

describe('the AudioTrackPublicationInfo component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <AudioTrackPublicationInfo publication={{ trackName: 'testName', trackSid: 'testSid' } as any} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render the AudioTrackInfo component if an audio track is not present', () => {
    mockUseTrack.mockImplementationOnce(() => null);
    const wrapper = shallow(
      <AudioTrackPublicationInfo publication={{ trackName: 'testName', trackSid: 'testSid' } as any} />
    );
    expect(wrapper.find(AudioTrackInfo).exists()).toBeFalsy();
  });
});
