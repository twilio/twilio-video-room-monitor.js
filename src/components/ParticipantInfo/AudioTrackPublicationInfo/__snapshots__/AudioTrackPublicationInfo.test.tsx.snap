// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the AudioTrackInfo component should render correctly if an audio track is present 1`] = `
<Fragment>
  <Memo(Datum)
    label="isSwitchedOff"
    value={false}
  />
  <Memo(Datum)
    label="Switch Off Reason"
    value="mock-switchoff-reason"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Jitter"
    value={2}
  />
  <Memo(Datum)
    label="Packets Lost"
    value={7}
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value={null}
  />
  <Memo(MediaStreamTrackInfo) />
</Fragment>
`;

exports[`the AudioTrackInfo component should render correctly when switchOffReason and isSwitchedOff are undefined (for local tracks) 1`] = `
<Fragment>
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Jitter"
    value={2}
  />
  <Memo(Datum)
    label="Packets Lost"
    value={7}
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value={null}
  />
  <Memo(MediaStreamTrackInfo) />
</Fragment>
`;

exports[`the AudioTrackPublicationInfo component should render correctly 1`] = `
<Memo(styled.div)>
  <Memo(Datum)
    label="Name"
    value="testName"
  />
  <Memo(Datum)
    label="SID"
    value="testSid"
  />
  <Memo(Datum)
    label="isSubscribed"
    value={true}
  />
  <AudioTrackInfo
    track="testTrack"
    trackSid="testSid"
  />
</Memo(styled.div)>
`;
